src/rust/inetstack/protocols/layer2/mod.rs

```diff
@@ -0,0 +1,149 @@
+// Associated Functions
+//======================================================================================================================
+
+impl SharedLayer2Endpoint {
+    pub fn new<P: PhysicalLayer>(config: &Config, layer1_endpoint: P) -> Result<Self, Fail> {
+        Ok(Self(SharedObject::new(Layer2Endpoint {
+            layer1_endpoint: Box::new(layer1_endpoint),
+            link_addr: config.local_link_addr()?,
+        })))
+    }
+
+    pub fn receive(&mut self) -> Result<ArrayVec<(Ethernet2Header, DemiBuffer), RECEIVE_BATCH_SIZE>, Fail> {
+        let mut batch: ArrayVec<(Ethernet2Header, DemiBuffer), RECEIVE_BATCH_SIZE> = ArrayVec::new();
+        for pkt in self.layer1_endpoint.receive()? {
+            let (header, payload) = match Ethernet2Header::parse(pkt) {
+                Ok(result) => result,
+                Err(e) => {
+                    // TODO: Collect dropped packet statistics.
+                    let cause: &str = "Invalid Ethernet header";
+                    warn!("{}: {:?}", cause, e);
+                    continue;
+                },
+            };
+            debug!("Engine received {:?}", header);
+            if self.link_addr != header.dst_addr()
+                && !header.dst_addr().is_broadcast()
+                && !header.dst_addr().is_multicast()
+            {
+                let cause: &str = "invalid link address";
+                warn!("dropping packet: {}", cause);
+            }
+            batch.push((header, payload))
+        }
+        Ok(batch)
+    }
+
+    pub fn transmit<P: PacketBuf>(&mut self, mut pkt: P) -> Result<(), Fail> {
+        let buf: DemiBuffer = match pkt.take_body() {
+            Some(body) => body,
+            _ => {
+                let cause = format!("No body in PacketBuf to transmit");
+                warn!("{}", cause);
+                return Err(Fail::new(libc::EINVAL, &cause));
+            },
+        };
+        self.layer1_endpoint.transmit(buf)
+    }
+}
+
+//======================================================================================================================
+// Trait Implementations
+//======================================================================================================================
+
+impl Deref for SharedLayer2Endpoint {
+    type Target = Layer2Endpoint;
+
+    fn deref(&self) -> &Self::Target {
+        self.0.deref()
+    }
+}
+
+impl DerefMut for SharedLayer2Endpoint {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.0.deref_mut()
+    }
+}
+
+/// Memory Runtime Trait Implementation for the network stack.
+impl MemoryRuntime for Layer2Endpoint {
+    /// Casts a [DPDKBuf] into an [demi_sgarray_t].
+    fn into_sgarray(&self, buf: DemiBuffer) -> Result<demi_sgarray_t, Fail> {
+        self.layer1_endpoint.into_sgarray(buf)
+    }
+
+    /// Allocates a [demi_sgarray_t].
+    fn sgaalloc(&self, size: usize) -> Result<demi_sgarray_t, Fail> {
```

Comment:
It would be good to annotate the variable name with the unit (e.g., `size_bytes` etc.) to avoid looking up at the flow of this variable in the code.

Reply:
I am only going to change it here. Since there is a trait, we would need to change this everywhere which would be a huge change.